### PR TITLE
Modified gemm, including transposed versions

### PR DESCRIPTION
**Pull request by abadams**
With this change we beat Eigen by a substantial amount for large
matrices (>2x on my machine). transA is the fast one (for us and for
eigen)

@kree-colemcalughlin PTAL

### CODE REVIEW

File: apps/linear_algebra/Makefile [hunk 1]
```diff
@@ -71,12 +77,14 @@ tests/test_halide_blas: tests/test_halide_blas.cpp $(LIBHALIDE_BLAS)
...
-L3_BENCHMARK_SIZES = 8 32 128 512
+# Large powers of two are a pathological case for the cache, so avoid
+# them for the benchmarks.
+L1_BENCHMARK_SIZES = 16 64 288 1056 2080
+L2_BENCHMARK_SIZES = 8 16 32 64 128 288 544 1056 2080
+L3_BENCHMARK_SIZES = 8 16 32 64 128 288 544 1056 2080
 L1_BENCHMARKS = scopy dcopy sscal dscal saxpy daxpy sdot ddot sasum dasum
 L2_BENCHMARKS = sgemv_notrans dgemv_notrans sgemv_trans dgemv_trans
-L3_BENCHMARKS = sgemm_notrans dgemm_notrans
+L3_BENCHMARKS = sgemm_notrans dgemm_notrans sgemm_trans dgemm_trans
```
Comment (kree-colemcalughlin) [inline apps/linear_algebra/Makefile]: This should be changed to sgemm_transA sgemm_transB etc...

File: apps/linear_algebra/src/blas_l3_generators.cpp [hunk 2]
```diff
@@ -42,64 +42,152 @@ class GEMMGenerator :
...
+
+        im_t.compute_root()
+            .specialize(rows >= 4 && cols >= 4)
+            .tile(i, j, ii, ji, 4, 4).vectorize(ii).unroll(ji)
+            .specialize(rows >= 128 && cols >= 128)
+            .tile(i, j, ti, tj, i, j, 16, 16)
+            .fuse(ti, tj, t).parallel(t);
+
+        transpose_tmp.compute_at(im_t, i)
+            .specialize(rows >= 4 && cols >= 4).vectorize(j).unroll(i);
```
Comment (kree-colemcalughlin) [inline apps/linear_algebra/src/blas_l3_generators.cpp]: Why is the vectorization size hardcoded to 4 here? Would it not be better to base this on natural_vector_size()? I guess maybe it all works out to the same thing after going through interleave_vectors.

**Comment by kree-colemcalughlin** [comment]
This is great news for the matrix multiplication work. I am working on updating the benchmarks comparing halide vs. other BLAS implementations. As you said, this shows that we can beat Eigen, but it is looking like we are getting close the other BLAS implementations as well. I would like to try compare with the non-pretransposed matrix-matrix multiply algorithm as well for completeness.

File: apps/linear_algebra/src/blas_l3_generators.cpp (hunk 2 above)
Comment (abadams) [inline apps/linear_algebra/src/blas_l3_generators.cpp]: It's the vector size for a in-register transpose, so the complexity of the code is more related to the number of things to transpose than their bit width. Four seemed good for floats and doubles.

**Comment by abadams** [comment]
Pushed some fixes and timing code changes. PTAL.

**Comment by kree-colemcalughlin** [comment]
Did you forget to commit benchmarks/macros.h? I don't see it after pulling your updates.

**Comment by abadams** [comment]
I did. Will fix tomorrow.

**Comment by abadams** [comment]
Fixed.

**Comment by kree-colemcalughlin** [comment]
LGTM. I like the work you did to pull the macros out. It greatly simplifies the code. I think we need to add some correctness testing for the transposed matrix multiplies. But let's get this merged and I can work that out in another PR along with selecting the algorithm based on problem size.

**Pull request merged**